transport/ssh.go

```diff
@@ -0,0 +1,180 @@
+package transport
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"time"
+
+	"github.com/kballard/go-shellquote"
+	"github.com/kevinburke/ssh_config"
+	"github.com/puppetlabs/wash/activity"
+	"github.com/puppetlabs/wash/datastore"
+	"github.com/puppetlabs/wash/plugin"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+	"golang.org/x/crypto/ssh/knownhosts"
+)
+
+// Cache SSH connections for better performance. Re-using SSH connections can significantly speed
+// up repeated SSH operations.
+var connectionCache = datastore.NewMemCacheWithEvicted(closeConnection)
+var expires = 15 * time.Second
+
+func closeConnection(id string, obj interface{}) {
+	if client, ok := obj.(*ssh.Client); ok {
+		client.Close()
+	}
+}
+
+func homeDir() (string, error) {
+	curUser, err := user.Current()
+	if err != nil {
+		return "", err
+	}
+
+	if curUser.HomeDir == "" {
+		return "", fmt.Errorf("the current user %v does not have a home directory", curUser.Name)
+	}
+	return curUser.HomeDir, nil
+}
+
+func newAgent() (ssh.AuthMethod, error) {
+	sshAgent, err := net.Dial("unix", os.Getenv("SSH_AUTH_SOCK"))
+
+	if err != nil {
+		return nil, err
+	}
+	return ssh.PublicKeysCallback(agent.NewClient(sshAgent).Signers), nil
+}
+
+func getHostKeyCallback() (ssh.HostKeyCallback, error) {
+	homedir, err := homeDir()
+	if err != nil {
+		return nil, err
+	}
+
+	return knownhosts.New(filepath.Join(homedir, ".ssh", "known_hosts"))
+}
+
+func sshConnect(host, port, user string) (*ssh.Client, error) {
+	connID := user + "@" + host + ":" + port
+	// This is a single-use cache, so pass in an empty category.
+	obj, err := connectionCache.GetOrUpdate("", connID, expires, true, func() (interface{}, error) {
+		agent, err := newAgent()
+		if err != nil {
+			return nil, fmt.Errorf("Failed to find config from SSH_AUTH_SOCK: %v", err)
+		}
+
+		hostKeyCallback, err := getHostKeyCallback()
+		if err != nil {
+			return nil, fmt.Errorf("Loading SSH known hosts file: %v", err)
+		}
+
+		sshConfig := &ssh.ClientConfig{
+			User:            user,
+			Auth:            []ssh.AuthMethod{agent},
+			HostKeyCallback: hostKeyCallback,
```

Comment:
I'm not sure it's worth reading known hosts yet, since it's very likely the fingerprint won't have been added for an arbitrary EC2 instance yet. I could also recommend adding
```
Host *.compute.amazonaws.com
  StrictHostKeyChecking no
```
and disable it that way.

Reply:
When would reading known hosts be bad/annoying? How likely are those scenarios?

Reply:
Ok, made that work.